zebra/dplane_fpm_nl.c

```diff
@@ -492,14 +492,130 @@ static void fpm_read(struct thread *t)
 		FPM_RECONNECT(fnc);
 		return;
 	}
-	stream_reset(fnc->ibuf);
+
+	/* Schedule the next read */
+	thread_add_read(fnc->fthread->master, fpm_read, fnc, fnc->socket,
+			&fnc->t_read);
+
+	/* We've got an interruption. */
+	if (rv == -2)
+		return;
+
 
 	/* Account all bytes read. */
 	atomic_fetch_add_explicit(&fnc->counters.bytes_read, rv,
 				  memory_order_relaxed);
 
-	thread_add_read(fnc->fthread->master, fpm_read, fnc, fnc->socket,
-			&fnc->t_read);
+	while (available_bytes) {
```

Comment:
Last change and everything will look OK: `available_bytes` must be read from `STREAM_READABLE` instead of the `stream_read_try` return.

If we had bytes read before `stream_read_try` will only return the new bytes count and not the currently available for read byte count.

Reply:
fixed